python/python/glide/async_commands/core.py

```diff
@@ -1232,3 +1234,104 @@ async def zrem(
             int,
             await self._execute_command(RequestType.Zrem, [key] + members),
         )
+
+    @dataclass
+    class PubSubMsg:
+        """ Describes incomming message """
+        message: str
+        """
+        Incomming message
+        """
+        channel: str
+        """
+        Name of an channel that triggered the message
+        """
+        pattern: Optional[str]
+        """
+        Pattern that triggered the message
+        """
+
+    class ChannelModes(Enum):
+        """ Describes subsciption modes. """
+        Exact = 1
+        """ Use exact channel names """
+        Pattern = 2
+        """ Use channel name patterns """
+        Sharded = 3
+        """ Use sharded PUBSUB. See  https://redis.io/docs/interact/pubsub/#sharded-pubsub for more details """
+
+    async def subscribe(self,
+                        channels_or_patterns: Set[str],
+                        channel_mode: ChannelModes,
+                        callback: Callable[[PubSubMsg, Optional[Any]]],
+                        context: Optional[Any]) -> None:
+        """
+        Add specific channels or set of channels defined by a pattern to active subscriptions.
+        The callback is associated with each channel/pattern,
+        meaning subsequent subsciption for the same channel/pattern will override the previous association, dereferencing the original callback object
+        Note1: Overlapping subscriptions produced by combinations of Exact and Pattern modes
+        count as distinc subscriptions and will produce duplicated messages.
+        Note2: Patterns are not applicable in Sharded mode. 
+        See https://redis.io/docs/interact/pubsub for more details.
+
+        Args:
+            channels_or_patterns: Set of channels or patterns to subscibe to.
+            channel_mode: Mode of operation. See ChannelModes
+            callback: Callback to be async-triggered for each new message accompanied with a context object. For each channel or pattern there is exactly one active callback.
+                    Subsequent calls to subscribe() will reset callbacks for existing channels or patterns.
+            context: User-provided context for this subscription        
+
+        Examples:
+            >>> await client.subscribe({"local-news"}, ChannelModes.Exact, process_news_message, None)
+                1  # Subscribes to local-news channel with process_news_message callback function
+            >>> await client.subscribe({"*"}, ChannelModes.Pattern, process_all_messages, None)
+                2  # Subscribes to all channels with process_all_message callback function
+        """
+        pass
+
+    async def publish(self,
+                      message: str,
+                      channels: Set[str],
+                      sharded: bool = False) -> int:
```

Comment:
why is the API different here? with sharded as a bool and all

Reply:
shouldn't we review this file?

Reply:
I am not following - we are reviewing this file :) Different from what

Reply:
oh, my bad. publish can't be done with patterns. this is why the api is different from subscribe :)

Reply:
done